[IV-23-24] Objetivo 8

adelahera opened the pull request:
## Issues
* Closes #68 
* Closes #70 
* Closes #71 
* Closes #72 
* Closes #73 
* Closes #74 

## Lista de comprobación

* [x] ¿Has justificado tu elección de un framework REST y explicado por qué no has
   elegido ni `flask` ni `express`?
* [x] ¿Las rutas están organizadas alrededor de un recurso?
* [x] ¿Has usado cada verbo HTTP con su forma convencional más aceptada?

adelahera committed:
Añade doc framework API, closes #68

Se documentan las diferentes opciones, se establecen los criterios y se explica la decisión final

adelahera committed:
Añade dependencias Fiber, closes #70

adelahera committed:
Añade diseño de rutas para la API, closes #71

Se han creado rutas para acceder a recursos como los jugadores, sus estadísticas y las épocas

adelahera committed:
Actualiza iv.yaml, closes #72

Se añade la nueva clave framework al iv.yaml

adelahera committed:
Crea nuevo package API, closes #73

Soluciona también la no superación de los tests, pues tenía que especificar el paquete del nuevo fichero

adelahera committed:
Actualiza metodos de la API, closes #74

Se añaden returns genéricos a las funciones para poder superar los tests

pluque01 commented on internal/api/api.go:
```diff
@@ -0,0 +1,61 @@
+	app.Get("/similares/:jugador/:temporada:/:similitud:/:estadistica", getSimilares)
+
+	// Borra un jugador
+	app.Delete("/borrar/:temporada/:jugador", deleteJugador)
```
No creo que estés orientando las rutas sobre el recurso. Las rutas deberían crearse alrededor del recurso y usar los verbos de HTTP para gestionar que tipo de acción se debe realizar. En este caso podría ser `app.Delete("/:temporada/:jugador", deleteJugador)`

adelahera replied:
Sí, había malentendido ese concepto. Arreglado en [7deb563](https://github.com/adelahera/basket-stats/pull/69/commits/7deb56380f8fd5f1861dcc02d8931c42a1407b63)

pluque01 commented on internal/api/api.go:
```diff
@@ -0,0 +1,61 @@
+	app.Delete("/borrar/:temporada/:jugador", deleteJugador)
+
+	// Crea un nuevo jugador
+	app.Post("/crear/:temporada/:jugador", addJugador)
+
+	// Actualiza los datos de un jugador
+	app.Put("/actualizar/:temporada/:jugador", updateJugador)
+
+	// Borra todos los jugadores de una época
+	app.Delete("/borrar/:epoca", deleteEpoca)
```
En estos lo mismo que te he comentado antes.

adelahera replied:
Arreglado en [7deb563](https://github.com/adelahera/basket-stats/pull/69/commits/7deb56380f8fd5f1861dcc02d8931c42a1407b63)

pluque01 reviewed (changes requested):
Hola!
Te he dejado una sugerencia, míralo cuando puedas :))

adelahera committed:
Actualiza rutas de la API, closes #71

Se rediseñan las rutas, orientando las rutas sobre el recurso al que acceden

pluque01 commented on internal/api/api.go:
```diff
@@ -0,0 +1,75 @@
+	app.Get("/jugador/:temporada:/:nombre/:estadistica/:similitud/:epoca", getSimilares)
+
+	// Borra un jugador de una temporada concreta
+	app.Delete("/jugador/:temporada/:nombre", deleteJugadorTemporada)
```
Yo cambiaría esta ruta para que fuera `/jugador/:nombre/:temporada`, así tendría más sentido con la siguiente.

pluque01 replied:
O incluso todas en las que haya temporadas poner el nombre primero

adelahera replied:
Gracias por la sugerencia. Arreglado en [a2e9f9c](https://github.com/adelahera/basket-stats/pull/69/commits/a2e9f9c5be0256601881deccb84e49aa695535de)

pluque01 reviewed (approved):
Yo lo veo bien, te he hecho una pequeña sugerencia pero te doy el visto bueno

adelahera committed:
Actualiza rutas de la API, closes #71

Se cambia el orden de los parametros para mayor comprension y coherencia

adelahera commented:
@JJ, puede echarle un vistazo cuando pueda? Muchas gracias 😃

JJ commented on internal/api/api.go:
```diff
@@ -0,0 +1,75 @@
+	app.Put("/estadisticas/:nombre/:temporada", updateEstadisticas)
+
+	// Borra las estadísticas de un jugador en una temporada
+	app.Delete("/estadisticas/:nombre/:temporada", deleteEstadisticas)
```
¿Qué necesidad hay de borrarlas?

adelahera replied:
En esto estoy de acuerdo, quizá no sea necesario. Arreglado en [91cec83](https://github.com/adelahera/basket-stats/pull/69/commits/91cec830dd66e9a088ba732f3705397e8c253a54)

JJ commented on internal/api/api.go:
```diff
@@ -0,0 +1,75 @@
+	app.Delete("/jugador/:nombre", deleteJugador)
+
+	// Crea un nuevo jugador
+	app.Post("/jugador", addJugador)
```
Tendrías que definir cuál es el URI de cada jugador, o de cada recurso. Si, como parece, es jugador/:nombre/:temporada, ya conoces el URI y por tanto ¿no sería mejor usar PUT en vez de POST?

adelahera replied:
Tuve esa misma duda mientras diseñaba las URIs. Al final me decanté por usar POST precisamente porque no conozco la URI del jugador, por eso la ruta es solo /jugador , ya que del jugador que se va a crear no conozco todavía ni su nombre ni la temporada. Es algo que me generó dudas en su momento, así que quizá tenga razón, pero no lo llego a entender del todo.

JJ replied:
¿Cómo puede ser que no lo conozcas? ¿Con qué parámetros lo creas entonces?

adelahera replied:
Pues los parámetros parar crear al jugador irían en el body, ¿o es incorrecto de esa forma?

JJ replied:
Los parámetros podrán ir en el body, pero lo que forme parte del URI (que parece ser el nombre y el periodo) es mejor que sean parte del mismo URI. Puedes usar PUT también, como siempre que se conoce el URI.

adelahera replied:
Una última duda en ese caso. Si utilizo PUT y el URI /jugador/:nombre/:temporada, sería la misma ruta que utilizo para actualizar los datos del jugador (la última ruta, que se encarga de actualizar datos de un jugador) ¿Cómo diferencio una ruta de la otra?

JJ replied:
Si se actualiza parcialmente debes usar `PATCH`, si lo actualizas completo, simplemente usa PUT de nuevo.

adelahera replied:
Entendido. Arreglado en [419a721](https://github.com/adelahera/basket-stats/pull/69/commits/419a721881a2cefabd4b8c2b8e0cb3dc7b2a70cc)

JJ reviewed (changes requested):
Bastante completo, pero convendría que le dieras una vuelta al diseño de los URIs

adelahera committed:
Actualiza metodos de la API, closes #71

Se elimina el metodo para borrar las estadisticas, al no ser necesario

adelahera committed:
Actualiza metodos de la API, closes #71

Se cambia de PUT a PATCH, al actualizar solo parcialmente, y de POST a PUT para crear un nuevo jugador

adelahera commented:
@JJ , entonces... ¿el objetivo estaría superado?

JJ commented:
Me temo que no, el estado es todavía "suggested changes...". ¿Estás pidiendo revisión?

adelahera commented:
> Me temo que no, el estado es todavía "suggested changes...". ¿Estás pidiendo revisión?

Efectivamente, se trataba de la revisión. Lo hago explícitamente.

JJ reviewed (approved):
:+1:
